sklearn/metrics/_dist_metrics.pyx

```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+    to :class:`sklearn.metrics.DistanceMetric` based on the physical
+    representation of the vectors (sparse vs. dense). It makes use of
+    cython.final to remove the overhead of method calls' dispatch.
+
+    Parameters
+    ----------
+    distance_metric: DistanceMetric
+        The distance metric responsible for computing distances
+        between two vectors of (X, Y).
+    """
+
+    @classmethod
+    def get_for(
+        cls,
+        X,
+        Y,
+        str metric="euclidean",
+        dict metric_kwargs=None,
+    ) -> DatasetsPair:
+        """Return the DatasetsPair implementation for the given arguments.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
+            Input data.
+            If provided as a ndarray, it must be C-contiguous.
+            If provided as a sparse matrix, it must be in CSR format.
+
+        Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features)
+            Input data.
+            If provided as a ndarray, it must be C-contiguous.
+            If provided as a sparse matrix, it must be in CSR format.
+
+        metric : str, default='euclidean'
+            The distance metric to use for argkmin. The default metric is
+            a fast implementation of the standard Euclidean metric.
+            For a list of available metrics, see the documentation of
+            :class:`~sklearn.metrics.DistanceMetric`.
+
+        metric_kwargs : dict, default=None
+            Keyword arguments to pass to specified metric function.
+
+        Returns
+        -------
+        datasets_pair: DatasetsPair
+            The suited DatasetsPair implementation.
+        """
+        cdef:
+            DistanceMetric distance_metric = DistanceMetric.get_metric(
+                metric,
+                **(metric_kwargs or {})
+            )
+
+        if X.dtype != np.float64 or Y.dtype != np.float64:
+            raise ValueError("Only 64bit float datasets are supported for X and Y.")
+
+        # Metric-specific checks that do not replace nor duplicate `check_array`.
+        distance_metric._validate_data(X)
+        distance_metric._validate_data(Y)
+
+        if issparse(X) or issparse(Y):
+            raise ValueError("Only dense datasets are supported for X and Y.")
+
+        return DenseDenseDatasetsPair(X, Y, distance_metric)
+
+    @classmethod
+    def unpack_csr_matrix(cls, X: csr_matrix):
+        """Ensure getting ITYPE instead of int internally used for CSR matrices."""
+        X_data = np.asarray(X.data, dtype=DTYPE)
+        X_indices = np.asarray(X.indices, dtype=ITYPE)
+        X_indptr = np.asarray(X.indptr, dtype=ITYPE)
+        return X_data, X_indptr, X_indptr
+
+    def __init__(self, DistanceMetric distance_metric):
+        self.distance_metric = distance_metric
+
+    cdef ITYPE_t n_samples_X(self) nogil:
+        """Number of samples in X."""
+        return -999
```

Comment:
That's a good remark. I think `raising` might take the GIL which is something that we do not want to happen.

To be honest I would favor being conservative on legacy (yet super efficient and clean) Cython code by not changing it if is works fine (which is the case here) and by keeping previous conventions.

I regret some cleaning I made to old but fine Cython code which slightly have change some users' code logic in rare and unexpected cases.

Reply:
> According to https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html?highlight=raise#releasing-the-gil, we may raise in Cython 3.0.

Cython 3 is still under development. We build scikit-learn with older versions (0.28+) so, in the mean time, we would need to explicitly acquire the GIL to raise an exception here:

```cython

    cdef ITYPE_t n_samples_X(self) nogil:
        """Number of samples in X."""
        with gil:
            raise TypeError("n_samples_X should be overriden in a concrete sublcass")

```

> That's a good remark. I think raising might take the GIL which is something that we do not want to happen.

I believe it's fine to acquire the GIL in a `nogil` section just to raise an exception: as soon as an exception happens here, we do not care about multi-threading performance anymore because any threaded code calling this will terminate to raise the exception anyway.

If the code is called with argument values that never cause the exception to be raised, the GIL will never be acquired and the multithreaded performance will stay optimal.

> To be honest I would favor being conservative on legacy (yet super efficient and clean) Cython code by not changing it if is works fine (which is the case here) and by keeping previous conventions.

But this is not legacy code here, right? The `DatasetsPair` and its subclasses are newly introduced in this PR so we do not have to worry about backward compat here. I would personally but in favor of using more modern Cython code style for new classes instead of replicating idioms of code that was written under the constraints of an 10 year old Cython version that we do not use anymore.

Reply:
> But this is not legacy code here, right? The DatasetsPair and its subclasses are newly introduced in this PR so we do not have to worry about backward compat here. I would personally but in favor of using more modern Cython code style for new classes instead of replicating idioms of code that was written under the constraints of an 10 year old Cython version that we do not use anymore.

I should have been clearer: this is not legacy code but I kept the legacy code convention for the sake of uniformity.

To me, updating the idiom is orthogonal to this PR's goal. I would favor doing a `MAINT` PR to address those small changes across the code base, if it's relevant (to me this does not look urgent at all).

Reply:
> To me, updating the idiom is orthogonal to this PR's goal.

I don't suggest to update the existing code as part of this PR. I just want to avoid propagating confusing idioms to newly written code when its easy to do so.

> I should have been clearer: this is not legacy code but I kept the legacy code convention for the sake of uniformity.

Code consistency is desirable in general but here I have the feeling that returning negative values when those methods are not overridden in subclasses can cause weird indirect errors with confusing error messages in `is_usable_for` or even possibly a segfault in this case. I think raising exceptions with explicit error messages could help make the code more maintainable.

I tried the following change locally, and it builds without error and tests pass:

```diff
diff --git a/sklearn/metrics/_dist_metrics.pxd b/sklearn/metrics/_dist_metrics.pxd
index e7c2f2ea2..13d3c9bd2 100644
--- a/sklearn/metrics/_dist_metrics.pxd
+++ b/sklearn/metrics/_dist_metrics.pxd
@@ -71,11 +71,11 @@ cdef class DistanceMetric:
 cdef class DatasetsPair:
     cdef DistanceMetric distance_metric
 
-    cdef ITYPE_t n_samples_X(self) nogil
+    cdef ITYPE_t n_samples_X(self) nogil except -1
 
-    cdef ITYPE_t n_samples_Y(self) nogil
+    cdef ITYPE_t n_samples_Y(self) nogil except -1
 
-    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil
+    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil except -1
 
     cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) nogil
 
diff --git a/sklearn/metrics/_dist_metrics.pyx b/sklearn/metrics/_dist_metrics.pyx
index 3def08da7..47d1ee451 100644
--- a/sklearn/metrics/_dist_metrics.pyx
+++ b/sklearn/metrics/_dist_metrics.pyx
@@ -1263,19 +1263,28 @@ cdef class DatasetsPair:
     def __init__(self, DistanceMetric distance_metric):
         self.distance_metric = distance_metric
 
-    cdef ITYPE_t n_samples_X(self) nogil:
+    cdef ITYPE_t n_samples_X(self) nogil except -1:
         """Number of samples in X."""
-        return -999
+        with gil:
+            raise NotImplementedError(
+                "n_samples_X should be overriden in a concrete sublcass"
+            )
 
-    cdef ITYPE_t n_samples_Y(self) nogil:
+    cdef ITYPE_t n_samples_Y(self) nogil except -1:
         """Number of samples in Y."""
-        return -999
+        with gil:
+            raise NotImplementedError(
+                "n_samples_Y should be overriden in a concrete sublcass"
+            )
 
     cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) nogil:
         return self.dist(i, j)
 
-    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil:
-        return -1
+    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil except -1:
+        with gil:
+            raise NotImplementedError(
+                "dist should be overriden in a concrete sublcass"
+            )
 
 @final
 cdef class DenseDenseDatasetsPair(DatasetsPair):
@@ -1302,11 +1311,11 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
         self.d = X.shape[1]
 
     @final
-    cdef ITYPE_t n_samples_X(self) nogil:
+    cdef ITYPE_t n_samples_X(self) nogil except -1:
         return self.X.shape[0]
 
     @final
-    cdef ITYPE_t n_samples_Y(self) nogil:
+    cdef ITYPE_t n_samples_Y(self) nogil except -1:
         return self.Y.shape[0]
 
     @final
@@ -1316,7 +1325,7 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
                                           self.d)
 
     @final
-    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil:
+    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) nogil except -1:
         return self.distance_metric.dist(&self.X[i, 0],
                                          &self.Y[j, 0],
                                          self.d)

```

While it works as expected, I suspect that the `except -1` in `dist` will add an additional conditional check when calling `dist` and that can possibly introduce a small slowdown when `return_distances=True`. I did not add the `except -1` on surrogate because I expect that some surrogate dist could be negative and that would cause weird problems.

So we could do the `except -1` only for `n_samples_X` / `n_samples_Y` but then the code would not be consistent. So all in all I am no strong opinion. I am fine with leaving the code the way it is.

Reply:
I checked and build-time checks for abstract base class do not exist in Cython. That would have been ideal.

Reply:
@lorentzenchr shall with resolve this conversation?

Reply:
Thank you for exploring this, @ogrisel.

I think the best is to keep the code convention consistent now and then update relevant Cython code for unimplemented methods for explicitness and safety in another PR. (it might be worth studying it in details as you started). What do you think?

Reply:
Marking this is resolved.